### PR TITLE
Name resolution load balancing

### PR DESCRIPTION
**Pull request by murgatroid99**
This adds a custom name resolver and load balancer internal API with basic implementations. The name resolver implementation handles retrieving and service config data from DNS TXT records, but most of the service config options are not implemented. The only implemented load balancer implementation is "pick first". It will attempt to establish connections to every address returned by the resolver, and will use whichever one connects first.

This change fixes the following bugs:

 - Calls on channels that quickly fail to establish a connection will end with an `UNAVAILABLE` error instead of waiting forever or until the call deadline. The `waitForReady` option to override that behavior is now supported. (#896)

 - Channels with no calls no longer keep the event loop alive. (#767)

Some parts of this design are not in this PR and will be added in future PRs:

 - UDS resolver
 - Round robin load balancer
 - Channel argument handling for the features implemented here

### CODE REVIEW

File: packages/grpc-js/src/channel.ts [hunk 1]
```diff
@@ -128,240 +105,220 @@ export interface Channel {
...
+        this.updateState(connectivityState);
+      },
+      requestReresolution: () => {
+        // This should never be called.
+        throw new Error(
+          'Resolving load balancer should never call requestReresolution'
+        );
+      },
+    };
+    // TODO: check channel arg for default service config
```
Comment (nicolasnoble) [inline packages/grpc-js/src/channel.ts]: Add your username to the TODO

File: packages/grpc-js/src/compression-filter.ts [hunk 2]
```diff
@@ -138,7 +138,9 @@ class UnknownHandler extends CompressionHandler {
   compressMessage(message: Buffer): Promise<Buffer> {
     return Promise.reject<Buffer>(
       new Error(
-        `Received message compressed wth unsupported compression method ${this.compressionName}`
+        `Received message compressed wth unsupported compression method ${
```
Comment (nicolasnoble) [inline packages/grpc-js/src/compression-filter.ts]: That's a bit of a weird looking change. What's the reasoning?

File: packages/grpc-js/src/index.ts [hunk 3]
```diff
@@ -283,3 +283,13 @@ export const InterceptingCall = () => {
 };
 
 export { GrpcObject } from './make-client';
+
+import * as resolver from './resolver';
+import * as load_balancer from './load-balancer';
+
+function setup() {
```
Comment (nicolasnoble) [inline packages/grpc-js/src/index.ts]: The typical paradigm for these is rather something along these lines:

```JS
(function() {
  resolver.registerAll();
  load_balancer.registerAll();
}());
```

File: packages/grpc-js/src/compression-filter.ts (hunk 2 above)
Comment (murgatroid99) [reply packages/grpc-js/src/compression-filter.ts]: I think this came from the code formatter.

Comment (paambaati) [reply packages/grpc-js/src/compression-filter.ts]: There’s a typo in `wth` — it should be `with`.

**Comment by felipegcampos** [comment]
Good job on these bugs. I think this PR might fix 2 problems we're currently experiencing. Do you have any estimations on when it should be released?

**Comment by murgatroid99** [comment]
@felipegcampos We want to get it out soon, but it's a lot of code so it's going to take some time for it to be reviewed.

**Comment by felipegcampos** [comment]
> @felipegcampos We want to get it out soon, but it's a lot of code so it's going to take some time for it to be reviewed.

Perfect. I was reviewing myself. However, I don't know if I'm of much help since I have not been checking the codebase since we started with grpc. However, let me know if I can help somehow.

**Pull request merged**